### ISSUE TITLE
Add option to skip sorting QuickPick items

### ISSUE THREAD

**Issue by pelmers** [microsoft/vscode#73904]
Similar to https://github.com/microsoft/vscode/issues/63050, I wish to avoid the final sort on the quick pick list. I am using the quickpick as a search field, and the `items` field I set is already sorted by my relevance metric, so I do not want to have VS Code sort the list again . Specifically, I want an option to skip this block: https://github.com/microsoft/vscode/blob/a69d0f8b218498e22b7c3a0f76926cc6322e6106/src/vs/workbench/browser/parts/quickinput/quickInputList.ts#L518-L523)

**Comment by pelmers** [comment]
I would like to implement this by adding a property `sortByValue` to `QuickPickOptions` on the extension API (`true` by default) to control this sorting. Does that sound reasonable and would you review such a PR? @chrmarti

**Comment by chrmarti** [comment]
I wonder if that would be sufficient. What about the highlighting of the search term on the labels? Would you want to turn that off too?

**Comment by pelmers** [comment]
@chrmarti highlighting the search term's not in the scope of this issue.

I am concerned the sorting case in particular because this sort() call happens in the renderer itself. Since the extension API expects a list of items rather than a set or map, it seems logical to me to let the extension to decide that this list is already sorted and let the renderer save some processing time (especially for large lists).

**Comment by chrmarti** [comment]
I don't want to add API without understanding the long-term direction and the highlighting, filtering and sorting all play into this.

The QuickPickOptions already have matchOnDescription and matchOnDetail flags and these control filtering and highlighting.

**Comment by pelmers** [comment]
Thanks for responding! I also noticed that filter/highlight can only be toggled for the description and detail fields, not the label field.

I imagine there was some kind of discussion when the API was originally added about its long-term vision; unfortunately that was before ['Hello Code'](https://github.com/microsoft/vscode/commit/8f35cc4768393b25468416829e980d7550619fb1) so it isn't in public history.

Allowing the extension to decide whether to sort results will let it provide a better experience to the user any time the default metric of 'index of first match' is not the best one.

For another example, an extension might want to keep the list grouped by category, similar to how built-in file search quickpick groups Recently opened, Search results, and Document symbols in that order.

**Comment by chrmarti** [comment]
We could add `matchOnLabel` (like we internally already did) to the API and couple that with filtering, highlighting and sorting. (Currently the sorting is only done based on the label.) That would fit nicely with the existing API while not exposing too fine-grained options that are hard to use.

**Comment by sergei-dyshel** [comment]
After reading the discussion I still fail to understand how we went from optional sorting feature request to discussing filtering options which are not dependent on sorting in any way....

**Comment by chrmarti** [comment]
When you want to control sorting yourself, you probably also want to control filtering. I don't want to introduce a flag specific to sorting now only to discover that doesn't solve the entire issue.

**Comment by pelmers** [comment]
Combining filtering/sorting into a single flag seems reasonable to me: if an extension implements its own filter/sort, it can just set all the `match` properties to false.

However highlighting probably can't be merged into the same flag, unless an extension could provide its own highlight ranges. It would be unfortunate if turning off sorting also requires giving up highlighting of the search label in the results.

**Comment by roblourens** [comment]
I agree that we should at least consider these things together even if we don't implement everything now. It's all about who controls the presentation of items in the quick pick. Also for example, if an extension has filtered and sorted based on its own algorithm, and vscode adds highlights based on a different algorithm, that might not make sense, so I could imagine a reason in the future to let extensions return highlight ranges too.

**Comment by chrmarti** [comment]
There is also the difficulty of the existing `matchOnDescription` and `matchOnDetail` flags controlling filtering and highlighting together. If we add `matchOnLabel` it should behave in a similar way (and maybe control sorting in addition).

**Comment by pelmers** [comment]
@chrmarti another option you could consider that would suit my requirements is adding an optional `sortText` field to the quick pick items, which has precedent in the completion items API:

https://github.com/microsoft/vscode/blob/fe1258b594c72c1ee014b66ee3193acf67d03f1d/src/vs/vscode.d.ts#L3329-L3334

**Comment by letmaik** [comment]
My scenario is that I'm querying a web service based on the filter text which returns the first 10 results. I would like to add a final "Load more results..." item (alternatively a button) which loads the next batch. However, the automatic sorting prevents me from doing that as the next 10 results wouldn't appear below the first 10, but in a more or less random order, which would be confusing.

My ask: an option to disable sorting but still have the highlighting enabled, together with using `alwaysShow: true` on all items to prevent hiding results based on the filter text.

**Comment by chrmarti** [comment]
@perlmers Could you explain the scenario you need to control sorting for? Why is the built-in sorting not sufficient? We are having offline discussions whether or not we should allow that much control by the extensions over the UX.

**Comment by pelmers** [comment]
@chrmarti apologies on late reply, I never got a notification message on this thread. 

Here's some brief details to expand on the use case I outlined in the issue opener:
- QuickPick is being used to display search results
- The 'label' field contains the line of text which matches the search
- The details field contains the path/line number that contains that line
- My preferred sorting is by (path, line) so results from the same file are grouped together in ascending line order
- The issue is today VS Code re-sorts the list by looking at 'label' first and discards the existing ordering

**Comment by Linskeyd** [comment]
@chrmarti, any thoughts on @pelmers expanded details above? It'd be great to have an option to skip the default logic for ordering results, maintaining the existing ordering of the provided results.

**Comment by chrmarti** [comment]
@Linskeyd It's on my list, I just haven't made it there yet.

**Comment by chrmarti** [comment]
I'll merge the PR as proposed API with the flag for the `createQuickPick()` API, but will not add it to `showQuickPick()`. `showQuickPick()` wouldn't benefit as much from it because it cannot update the list once the UI is shown and there is also value in keeping that part of the API simple while `createQuickPick()` can cover the advanced cases.

Hope that makes sense. This will still need an API review before we can move it to the stable API.

**Comment by chrmarti** [comment]
I haven't seen any feedback on this. @pelmers Does this flag do what you need and are you using it?

**Comment by letmaik** [comment]
@chrmarti I have the same need and the new `sortByLabel` does the job for me.

**Comment by pelmers** [comment]
@chrmarti thanks for following up, we've been using the flag as well and no longer see complaints on ordering.

**Comment by friedrb** [comment]
What about providing an option that simply turns off all filtering, sorting, and highlighting so that developers can completely handle the contents of the items themselves.  That would let advanced developers do what they want in response to the search text changing, yet still provide a default behavior for less complex needs.  For example, I would like to implement multiple search terms separated by spaces (like ctrl-shift-p has). But the default filtering doesn’t allow this.

**Comment by rmanthorpe** [comment]
@friedrb I agree this is definitely a missing feature. See https://github.com/microsoft/vscode/issues/83424 and https://github.com/microsoft/vscode/issues/83425 which between them should achieve what you want.

**Comment by friedrb** [comment]
I'd just like to add that for those using TypeScript, you can access QuickPick's new sortByLabel flag like so:
(Since it's not in the type definitions yet.)
```
let quickPick = vscode.window.createQuickPick();
(quickPick as any).sortByLabel = false;
```

**Comment by TylerLeonhardt** [comment]
Is this fixed now that there's a `sortByLabel` property?

**Comment by pelmers** [comment]
yes I think so, thanks!

**Comment by karrtikr** [comment]
> Is this fixed now that there's a sortByLabel property?

@pelmers @TylerLeonhardt Does it also work when we're currently filtering the list? IIRC it didn't used to work in that case.

**Comment by pelmers** [comment]
@karrtikr do you mean when you're typing in the list, like in https://github.com/microsoft/vscode/issues/54400 ?

**Comment by karrtikr** [comment]
@pelmers Yes, I would like to have the order untouched as probably mentioned here https://github.com/microsoft/vscode/issues/54400#issuecomment-405282034. That was closed in favor in this issue, so I expected this issue to provide such an option.

**Comment by pelmers** [comment]
@karrtikr I believe that if you turn off sortbylabel it will also keep your order when filtered from the input. But I'm not 100% sure since I don't work on this recently

**Comment by karrtikr** [comment]
It's not working for me, the order changes when filtering.

**Comment by pelmers** [comment]
ok since it sounds like we still can't skip sorting when the field has input, I'll reopen this issue.

However @TylerLeonhardt please note https://github.com/microsoft/vscode/issues/54400 is probably more specific to this case so maybe we should close this and reopen that instead.

**Comment by Brian-Bartels** [comment]
I'm not sure I understand why the filtering operation does any sorting at all. Is that an expected behavior of a filter?

**Comment by gjsjohnmurray** [comment]
@TylerLeonhardt can this `quickPickSortByLabel` proposed API get finalized soon? I recently used it in #199473 in the Git extension, which is built-in so doesn't have any problem using proposed APIs. But we extension developers can't use proposed APIs if we want to publish on Marketplace.

**Comment by lszomoru** [comment]
@TylerLeonhardt, would it be possible to also add this as an option to `window.showQuickPick`? Thanks!

**Comment by TylerLeonhardt** [comment]
This iteration, we are planning on removing the sort logic for the bigger `createQuickPick` API. That would mean no API would be needed, and if you want the existing sort behavior, you can just call `.sort`. I'll rename the issue accordingly.

**Comment by gjsjohnmurray** [comment]
@TylerLeonhardt I'm trying to understand how this won't be a breaking change.

**Comment by TylerLeonhardt** [comment]
Right, yeah it's a change we'd like to make to get a read on how many extensions depend on the existing behavior. Our suspicion is that this existing behavior is hurting extensions more than it helps and thus the behavior of the API should change.

Naturally, this is not an actual API breakage, since no API shape is changing... it's a behavior difference.

There is a chance that such a change would cause issues... if so, we'll re-evaluate.

**Comment by simonvanbernem** [comment]
Hi,

I'm new to developing for VSCode and am very interested in this API-change (to create a go-to symbol that is closer to the one from sublime text), but seeing that it got put back into the backlog after being part of milestone twice is confusing to me.

Does that mean it won't happen in the near future? Was it decided to not implement this change after all?

**Comment by calsys456** [comment]
I successfully avoided misleading sort by zipping zero-width spaces into the label, so that the QuickPickItem can never be matched and sorted, and i can sort & filter items manually myself. The only downside is that the incremental matched-highlight will missing, but i think it's not a big problem against the chaotic ordered items... Put it here for those who desperately want this feature lol

```typescript
class ManualSortItem implements vscode.QuickPickItem {
  label: string = "";
  origLabel: string = "";
  get labelText() {
    return this.origLabel;
  }
  set labelText(text: string) {
    // Listen to me you VSCode scum: you can root around all you want, but
    // you'll never match me, and you'll never sort me! You call yourself a
    // programmer?
    this.label = [...text].join("\u200B");
    this.origLabel = text;
  }
  constructor(label: string) {
    this.labelText = label;
  }
}
```
